homeassistant/components/radarr/sensor.py

```diff
@@ -95,154 +76,137 @@
         if self.entity_description.key == "status":
-            return self.data.attributes
-        return attributes
-
-    async def async_update(self) -> None:
-        """Update the data for the sensor."""
-        try:
-            if self.entity_description.key == "diskspace":
-                data = await self.radarr.async_get_diskspace()
-                if len(self.included) > 0:
-                    self.data = [datum for datum in data if datum.path in self.included]
-                else:
-                    self.data = data
-                space = sum(mount.freeSpace for mount in self.data)
-                self._attr_native_value = (
-                    f"{to_unit(space, self.native_unit_of_measurement):.2f}"
-                )
-            if self.entity_description.key == "upcoming":
-                local = dt_util.start_of_local_day().replace(microsecond=0)
-                start = dt_util.as_utc(local)
-                end = start + timedelta(days=self.days)
-                self.data = await self.radarr.async_get_calendar(
-                    start_date=start, end_date=end
-                )
-                self._attr_native_value = len(self.data)
-            if self.entity_description.key == "commands":
-                self.data = await self.radarr.async_get_commands()
-                self._attr_native_value = len(cast(list, self.data))
-            if self.entity_description.key == "status":
-                self.data = await self.radarr.async_get_system_status()
-                self._attr_native_value = self.data.version
-            if self.entity_description.key == "movies":
-                self.data = await self.radarr.async_get_movies()
-                self._attr_native_value = len(cast(list, self.data))
-        except ArrException as ex:
-            _LOGGER.warning(ex)
-        self._attr_available = self.native_value is not None
-
-
-def release_date(movie: RadarrMovie) -> str:
-    """Get release date."""
-    date = movie.physicalRelease or movie.digitalRelease or movie.inCinemas
-    return date.isoformat()
-
-
-def to_key(movie: RadarrMovie):
+            return self.coordinator.system_status.attributes
+        return {to_key(movie): movie.releaseDate for movie in self.coordinator.calendar}
+
+    @property
+    def native_value(self) -> StateType:
+        """Return the state of the sensor."""
+        if self.entity_description.key == "diskspace":
+            space = 0
+            for mount in self.coordinator.rootfolder:
+                space += mount.freeSpace
+            return f"{to_unit(space, cast(str, self.native_unit_of_measurement)):.2f}"
+        if self.entity_description.key == "upcoming":
+            return len(self.coordinator.calendar)
+        if self.entity_description.key == "commands":
+            return len(self.coordinator.commands)
+        if self.entity_description.key == "movies":
+            return len(self.coordinator.movies)
```

Comment:
I would add spport for the entity description to specify a `get_native_value`, it could work like `get_native_value=lambda coordinator: coordinator.moves`

Then this method would just do `return self.entity_description.get_native_value(self.coordinator)`

Reply:
I assume this would be done like `venstar`? It seems to mess with the way I have my dataclasses in `aiopyarr` where there it is `@dataclass(init=False)`

Reply:
Like Shelly https://github.com/home-assistant/core/blob/dev/homeassistant/components/shelly/sensor.py#L125

EntityDescriptions should live in Home Assistant as it's a way to describe the data for Home Assistant entities. Shouldn't tie to `aiopyarr`  ?